iree/compiler/Dialect/Flow/Transforms/DeduplicateExecutables.cpp

```diff
@@ -24,111 +25,227 @@ namespace Flow {
+          lhs.getBlocks(), rhs.getBlocks(),
+          [&](Block &lhsBlock, Block &rhsBlock) {
+            if (lhsBlock.getNumArguments() != rhsBlock.getNumArguments())
+              return false;
+            for (auto argPair :
+                 llvm::zip(lhsBlock.getArguments(), rhsBlock.getArguments())) {
+              auto &lhsArg = std::get<0>(argPair);
+              auto &rhsArg = std::get<1>(argPair);
+              if (lhsArg.getType() != rhsArg.getType()) return false;
+              mapping.map(lhsArg, rhsArg);
+            }
+            mapping.map(&lhsBlock, &rhsBlock);
+            return true;
+          })) {
+    return false;  // block mismatch
+  }
+
+  // Walk the blocks again now that we have a populated mapping.
+  for (auto blockPair : llvm::zip(lhs.getBlocks(), rhs.getBlocks())) {
     auto &lhsBlock = std::get<0>(blockPair);
     auto &rhsBlock = std::get<1>(blockPair);
-    // Warning: .size() is linear time.
-    // We could instead iterate through both lists of operations explicitly,
-    // stopping when operations are not equivalent, OR either list runs out of
-    // operations early.
-    if (lhsBlock.getOperations().size() != rhsBlock.getOperations().size()) {
-      return false;
-    }
-
     for (auto opPair :
          llvm::zip(lhsBlock.getOperations(), rhsBlock.getOperations())) {
       auto &lhsOp = std::get<0>(opPair);
       auto &rhsOp = std::get<1>(opPair);
-      if (!OperationEquivalence::isEquivalentTo(
-              &lhsOp, &rhsOp, OperationEquivalence::IgnoreOperands)) {
+      if (!isStructurallyEquivalentTo(lhsOp, rhsOp, mapping)) {
         return false;
       }
+    }
+  }
 
-      // We want to check the operand _types_, but don't care if the actual
-      // operand references differ (as they live in separate modules anyway).
-      if (!std::equal(lhsOp.operand_type_begin(), lhsOp.operand_type_end(),
-                      rhsOp.operand_type_begin())) {
-        return false;
-      }
+  // Equivalent!
+  return true;
+}
+static bool isStructurallyEquivalentTo(Operation &lhs, Operation &rhs,
+                                       BlockAndValueMapping &parentMapping) {
+  // Check operation metadata for early-exit opportunities.
+  if (lhs.getName() != rhs.getName()) return false;
+  if (lhs.getNumOperands() != rhs.getNumOperands()) return false;
+  if (lhs.getNumResults() != rhs.getNumResults()) return false;
+  if (lhs.getNumRegions() != rhs.getNumRegions()) return false;
+  if (lhs.getNumSuccessors() != rhs.getNumSuccessors()) return false;
+
+  // TODO(#3996): symbol mapping; for now allow them to differ unconditionally.
+  if (!compare_ranges(
+          lhs.getAttrs(), rhs.getAttrs(),
+          [&](const NamedAttribute &lhs, const NamedAttribute &rhs) {
+            if (lhs.first == "function_ref" ||
+                lhs.first == SymbolTable::getSymbolAttrName()) {
+              return true;
+            }
+            if (lhs.first != rhs.first) return false;
```

Comment:
I think NamedAttribute is just a std::pair, so lhs == rhs should just work I think.

Reply:
I meant just for the last couple lines

```
if (lhs.first != rhs.first) return false;
if (lhs.second != rhs.second) return false;
return true;
```

Can just be `return lhs == rhs`. Agreed that the special case for the symbol attributes above is still needed.